### ISSUE TITLE
pretrain 시 입력값 형태 질문

### ISSUE THREAD

**Issue by KimJaehee0725** [SKT-AI/KoBART#18]
안녕하세요! 우선 KoBART를 공개해주셔서 감사합니다! 덕분에 편하게 프로젝트를 진행할 수 있게 되었습니다!

친구들과 함께 KoBART를 이용하여 pegasus를 개발하려 하는데, KoBART pretrain 시 구체적인 입력값을 찾지 못해 질문드리게 되었습니다. 

물론 이전에도 https://github.com/SKT-AI/KoBART/issues/2 처럼 이슈를 제기해주신 분이 계신데, 제가 지식이 부족한 탓에 잘 이해가 가지 않아 다시 질문드리게 되었습니다. 

제가 이해한 바로는 다음과 같이 여러 문장을 인코더나 디코더에 싣고 계신데 혹시 맞나요?

- raw text: 이렇게 문장을 입력하면 되나요? 맞는지 궁금합니다.
- tokenized : 
![image](https://user-images.githubusercontent.com/43404665/145987879-81435a5f-669e-4ef9-8046-c901074c2419.png)

**Comment by bage79** [comment]
@KimJaehee0725 
알려주신 예제는 총 2문장이므로, 문장 단위로 분리하여 "이렇게 문장을 입력하면 되나요?" 와 "맞는지 궁금합니다."로 각각 입력을 만드셔야 합니다.
BART의 경우, 각 문장당 입력이 총 3개 존재하며, 예제로 적어주신 "이렇게 문장을 입력하면 되나요?"에 대하여 max_seq_len가 11 인 경우 아래와 같은 형태가 될 것입니다.
```text
encoder in: _이렇게 _문 장을 <mask> 하면 _되 나 요? </s> <pad> <pad>
decoder in: <s> _이렇게 _문 장을  _입력 하면 _되 나 요? </s> <pad> 
label:      _이렇게 _문 장을 _입력  하면 _되 나 요? </s> <pad> <pad>
```
label을 기준으로, encoder in 은 일부 토큰이 MASK 토큰으로 치환되었다고 생각하시면 되고,
decoder in 은 label 앞에 bos 토큰이 추가되었다고 생각하시면 될 것 같습니다.

BART논문과 동일하게, MASK 토큰은(max_seq_len * 0.15) 개수 만큼 선택하였습니다.

**Comment by KimJaehee0725** [comment]
우선 친절한 답변 감사드립니다! 
제가 착오가 있었네요
기존의 BART 모델과 비교하여 약간 다른 점이 있는 것 같은데 맞는지 궁금합니다.

1. 영어 Bart 모델의 경우엔 아래와 같이 encoder input으로 bos 토큰도 입력하는데, KoBart에선 사용하지 않나요?
![image](https://user-images.githubusercontent.com/43404665/146330976-783ab949-3401-46b1-afa9-71206ddfa062.png)

2. 아래와 같이 영어 Bart에선 decoder bos 토큰으로 eos 토큰을 사용하는데 KoBart는 bos를 사용한건가요?
https://github.com/huggingface/transformers/blob/master/src/transformers/models/bart/configuration_bart.py#L138

**Comment by bage79** [comment]
@KimJaehee0725 
먼저 답변이 늦어져서 죄송합니다.

1. 제가 이해한 것이 맞다면, BART encoder 에는 bos 토큰이 들어가지 않고, mask 토큰만 추가되는 것으로 알고 있습니다. decoder 에서는 물론 bos 토큰을 사용하구요.
![image](https://user-images.githubusercontent.com/304911/160570465-e6b8f201-f3cd-4e67-bf5b-2fc9d07201b3.png)

2. 아래 그림처럼 영어 tokenizer(BartTokenizer)에서 bos, eos 를 모두 사용하고 있구요. "decoder bos 토큰으로 eos 토큰을 사용하는데" 이 부분을 제가 잘 이해하지 못했습니다.
그리고 현재의 한글 tokenizer는 BartTokenizer를 상속받지 않고, 별도로 구현되어 post processing (template processing)이 포함되어 있지 않습니다. 따라서 영어 tokenizer 와 동작하는데 차이가 있음을 양해 부탁드립니다.
![image](https://user-images.githubusercontent.com/304911/160574726-0bb8cfe5-5ed6-4b9d-b70a-de32a5879257.png)

**Comment by ycahn82** [comment]
[bage79](https://github.com/bage79)님

안녕하세요. Kobart를 사용할 수 있도록 공유해주셔서 감사하다는 말씀을 먼저 드립니다..

저는 공유해 주신 Kobart를 활용하여 Open domain QA 모델을 만들고 있는데요. 이상한 점이 있어 질문 드립니다.

finetuning 단계에서 decoder의 입력으로 label을 위에서 말씀하신 것과 똑같은 형태로 아래와 같이 입력을 주면 성능이 잘 안나옵니다. 
(정확히 말하자면,  학습시 loss는 잘 떨어지는데 생성이 잘 안됩니다.  (학습을 완료하고 생성을 하면, \<s>\<s> tok1 tok2 와 같이 \<s>가 2번 나오게 되며 성능지표인 Exact match(EM) 스코어가 0.2 정도로 낮습니다.)
``` 
label                : tok1 tok2 tok3 </s> 
decoder input        : <s>  tok1 tok2 tok3 
```

반면,  아래와 같이 decoder input을 </s>로 시작하도록 바꾸어 주면 loss도 비슷하게 잘떨어지고, 생성도 잘되는 것 (0.4 정도로 좀 더 높습니다.) 을 확인할 수 있습니다. 
``` 
label                : tok1 tok2 tok3 <s> 
decoder input        : </s> tok1 tok2 tok3 
``` 
왜 이런 현상이 일어나는지 아시는 부분이 있으실지요?

혹시나 해서 github main화면의 README에 공유해 주신 코드를 따라 들어 가봤습니다.
Summarization  example에서는 \</s> 가 먼저 입력하도록 data loading 코드가 되어 있고, chitchat example에서는 \<s>를 먼저 입력하도록 되어 있더군요. 왜 이런 차이가 발생하는 지 궁금합니다.

혹시 아시는 부분이 있다면 알려주시면 감사하겠습니다.

**Comment by sanghyuk-choi** [comment]
제가 author이나 contributor은 아니지만, 이것 저것 테스트 해본 바를 댓글로 남깁니다.
BART논문 그림의 `<s>`를 huggingface와 연관 지어서 본다면, 그림의 `<s>`는 bos token이나 eos token을 의미하는게 아니라`decoder_input_ids` 를 뜻하는 것 같습니다.
즉 논문에 쓰인 `<s>` 는 seq2seq 모델에서 decoder 학습시 첫 token으로 들어가는 `<go>` token을 의미하는 것 같고요, 

이는 bos token이나 eos token과는 별개라고 보시면 될 것 같습니다. 
다만 huggingface bart 에서는 `<go>` token 으로 eos token을 사용하고요.
만약 bos token을 따로 붙여준다면, 그림에서 `A` token 이 bos token이 됩니다. 
**즉 huggingface bart 에서는 bos token이 `<go>` token이 아니라는 뜻 입니다.**

예를들어 eos token과 bos token을 다 사용하고 `<go>` token으로 eos token을 사용한다면
```
label                : <bos> tok1 tok2 tok3 <eos>
decoder input        : <eos> <bos> tok1 tok2 tok3
```
위와 같이 될 것입니다.

다만 huggingface 에서 bart 모델을 fine-tuning 할때 기본적으로 feed 하는것은 label 입니다. 
모델을 call 할 때 `<go>` token을 넣어서 feed 할 필요가 없이, 그냥 label을 feed 하면
모델 내부적으로 `prepare_decoder_input_ids_from_labels` 함수를 호출해서 decoder의 입력으로 `<go>` token을 맨 앞에 붙여줍니다. 

skt/kobart 과 gogamza/kobart-base-v2 가 같은지는 모르겠지만, 후자의 경우 config를 보면
encoder 쪽은 그냥 bos, eos token 둘다 붙이지 않고 **원문 텍스트 그대로 input**으로 들어가면 되고
decoder 쪽은 **원문 텍스트 + eos** 이렇게 feed 하면 잘 동작하는 것 같습니다.
(즉 bos token은 사용하지 않는것으로 보입니다.)

만약 skt/kobart 와 gogamza/kobart-base-v2 가 다르다면 skt/kobart의 bart config 및 tokenizer config를 확인하여서 사용하시면 될 것 같습니다.

**Comment by ycahn82** [comment]
답변 감사합니다. 명쾌하게 이해가 갔네요 ^^
